op/reboot.go

```diff
@@ -18,96 +18,336 @@ import (
+	o.failedNodes = append(o.failedNodes, node)
 	o.mu.Unlock()
 }
 
-// RebootOp returns an Operator to reboot nodes.
-func RebootOp(apiserver *cke.Node, nodes []*cke.Node, index int64, config *cke.Reboot) cke.InfoOperator {
-	return &rebootOp{
-		apiserver: apiserver,
-		nodes:     nodes,
-		index:     index,
-		config:    config,
+func (o *rebootDrainStartOp) Targets() []string {
+	ipAddresses := make([]string, len(o.entries))
+	for i, entry := range o.entries {
+		ipAddresses[i] = entry.Node
 	}
+	return ipAddresses
 }
 
-func (o *rebootOp) Name() string {
-	return "reboot"
+func (o *rebootDrainStartOp) Info() string {
+	if len(o.failedNodes) == 0 {
+		return ""
+	}
+	return fmt.Sprintf("failed to drain some nodes: %v", o.failedNodes)
 }
 
-func (o *rebootOp) NextCommand() cke.Commander {
-	switch o.step {
-	case 0:
-		o.step++
-		return rebootStartCommand{index: o.index}
-	case 1:
-		o.step++
-		nodeNames := make([]string, len(o.nodes))
-		for i := range o.nodes {
-			nodeNames[i] = o.nodes[i].Nodename()
-		}
-		return cordonCommand{
-			apiserver:     o.apiserver,
-			nodeNames:     nodeNames,
-			unschedulable: true,
-		}
-	case 2:
-		o.step++
-		return drainCommand{
-			timeoutSeconds:      o.config.EvictionTimeoutSeconds,
-			apiserver:           o.apiserver,
-			nodes:               o.nodes,
-			protectedNamespaces: o.config.ProtectedNamespaces,
+func (o *rebootDrainStartOp) NextCommand() cke.Commander {
+	if o.finished {
+		return nil
+	}
+	o.finished = true
+
+	return rebootDrainStartCommand{
+		entries:             o.entries,
+		protectedNamespaces: o.config.ProtectedNamespaces,
+		apiserver:           o.apiserver,
+		notifyFailedNode:    o.notifyFailedNode,
+	}
+}
+
+func (c rebootDrainStartCommand) Command() cke.Command {
+	ipAddresses := make([]string, len(c.entries))
+	for i, entry := range c.entries {
+		ipAddresses[i] = entry.Node
+	}
+	return cke.Command{
+		Name:   "rebootDrainStartCommand",
+		Target: strings.Join(ipAddresses, ","),
+	}
+}
+
+func (c rebootDrainStartCommand) Run(ctx context.Context, inf cke.Infrastructure, _ string) error {
+	cs, err := inf.K8sClient(ctx, c.apiserver)
+	if err != nil {
+		return err
+	}
+	nodesAPI := cs.CoreV1().Nodes()
+
+	protected, err := listProtectedNamespaces(ctx, cs, c.protectedNamespaces)
+	if err != nil {
+		return err
+	}
+
+	// Draining should be done sequentially.
+	// Parallel draining is relatively prone to deadlock.
+
+	// first, cordon all nodes
+	evictNodes := []*cke.RebootQueueEntry{}
+	for _, entry := range c.entries {
+		err := func() error {
+			entry.Status = cke.RebootStatusDraining
+			entry.LastTransitionTime = time.Now().Truncate(time.Second).UTC()
+			err = inf.Storage().UpdateRebootsEntry(ctx, entry)
+			if err != nil {
+				return err
+			}
+
+			err = checkJobPodNotExist(ctx, cs, entry.Node)
+			if err != nil {
+				return err
+			}
+
+			_, err = nodesAPI.Patch(ctx, entry.Node, types.StrategicMergePatchType, []byte(`
```

Comment:
If the check uses cached information, even the "cordon and check" steps can overlook a new Job Pod.
As described above this oversight would be acceptable.
It would be good to comment about this.

The comment "This should not happen" in `checkPodDeletion()` seems too strong.

Reply:
There is a race as you pointed out, but I think it's acceptable. I will add a comment.

> The comment "This should not happen" in checkPodDeletion() seems too strong.

I agree.